api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	corev1 "k8s.io/api/core/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+//
+// ExternalDNS describes a managed ExternalDNS controller instance for a cluster.
+// The controller is responsible for creating external DNS records in supported
+// DNS providers based off of instances of select Kubernetes resources.
+type ExternalDNS struct {
```

Comment:
When I delete my instance, how are you managing the cleanup?  My expectation is a finalizer as you tidy up, but how can you distinguish between a dns record this operand creates and a DNS record someone manually created?

Reply:
How would finalizers apply in this scenario? ExternalDNS creates DNS records for resources in the cluster, such as services or routes. It wouldn't make sense for ExternalDNS to delete the cluster resources themselves. Arguably, it might make sense to additionally leave external DNS record cleanup to a cluster-administrator.

Reply:
It looks like it's technically feasible to find all the records owned by an instance of ExternalDNS.
Each DNS record created by ExternalDNS is accompanied by a corresponding TXT record which contains the ownerID (must be unique for each ExternalDNS instance).
It's possible to list all the records of the given zones and filtering on TXT records which would allow us to check the ownership.
I think that I would prefer to see this clean-up functionality in ExteranlDNS rather than in the operator due to the number of supported providers.

Reply:
Right, my concern was more so figuring out how to tell ExternalDNS to explicitly delete a set of records. Do you know if there is a way to do that without modifying the record's source resources?

Reply:
Without modifying the source - no, I didn't find anything for it.

Reply:
I don't see doc updates for this. But I think you need to
1. cleanup external-dns created DNS entries.  If this feature is missing now, it seem important to create rather than only having orphaning policy
2. the operator needs to clean up the external-dns controller itself.
it seems like a finalizer may be required.